include/xtensor/xshape.hpp

```diff
@@ -30,11 +30,15 @@ namespace xt
     template <class T, std::size_t N>
     using static_shape = std::array<T, N>;
 
+    template<typename T, std::size_t N>
+    xt::static_shape<std::size_t, N> shape(const T(&aList)[N]);
+
```

Comment:
The return type should be `xt::static_shape<T, N>`.
Nitpicking: the declaration should be below the `constexpr` generic overload

Reply:
Thanks for pointing this out! Please review latest update.

Reply:
I was actually wrong, changing `std::size_t` for `T` does not work because the constructors expect `std::array<std::size_t, N>` while the initialisier list is usually `int` (like in `{ 1, 17, 42 }`). So it was expected that different types would be cast by the `std::copy()`. Is that acceptable?

Reply:
So the idea would be to use it this way `auto s = shape<std::size_t>({1, 17, 42});`, however this is cumbersome to write and I'm not even sure that this would compile.

The following would be better:
```cpp
template <class R = std::size_t, class T, std::size_t N>
xt::static_shape<R, N> shape(const T(&aList)[N])
{// ....
}
```

This way T is correctly detected, and the default return type avoids the need to type `std::size_t` for each invocation.

The only possible issue is the warning emitted by `std::copy`, in that case `for_each` with a lmabda performing the cast and copy might be prefered.

Reply:
Now I see your point, thanks. I've updated the code again with this suggestion. Next week I may have a few minutes to extend it with a casting lambda.

Would you want to include a check that unsigned types are not negative before the cast to `size_t`, or is that something you avoid in your code?

Reply:
That's something we try to avoid; if you feel an additional check would be better, you can add it with the `XTENSOR_ASSERT` macro so it is compiled and effective only when xtensor assertions are enabled.